mdbook/src/aizuna_config.md
# Aizuna 設定
Aizunaの動作は設定ファイルによって制御されます。

## Aizuna 起動オプション
`aizuna --help`で起動オプションを確認できます。

```shell
$ aizuna --help

aizuna v0.1.0

Usage:
    aizuna [Options]

Options:
    -v, --version   print version
    -h, --help      print this help menu
    -R, --root PATH set Aizuna root path. default '${HOME}/.config/aizuna'
```

## Aizuna ルートディレクトリ
設定ファイルが保存されるディレクトリを差します。

デフォルトでは `{HOME}/.confog/aizuna/` です。

ルートディレクトリは、Aizuna起動オプション `--root / -R` で 変更できます。

```
aizuna --root ./
```

Aizunaの実行プロセスはルートディレクトリに対してファイルの作成権限を必要とします。

## 設定ファイル
ルートディレクトリに存在する`config.toml`という名前の、toml形式のテキストファイルです。

設定ファイルが存在しない場合、自動的に作成されます。


```toml
# -*- mode:toml; coding:utf-8-unix; -*-
# /////////////////////////////////////////////////////////////////////////////
# =============================================================================
# aizuna v0.1.0
serdever    = 0
greeting    = "aizuna v0.1.0"
driver      = "Thread"
fringe_stack_size   = 1048576
path_db     = "./db"
prefix      = ","
# /////////////////////////////////////////////////////////////////////////////
# =============================================================================
[connectors.console]
serdever    = 0
enable      = true
connector   = "console"
# =============================================================================
[connectors.discord-00]
serdever    = 0
enable      = false
connector   = "discord"
[connectors.discord-00.config]
serdever    = 0
token       = "DISCORD_BOT_TOKEN"
# /////////////////////////////////////////////////////////////////////////////
# =============================================================================
[rules.shinen]
serdever    = 0
enable      = false
prefix      = ","
[rules.shinen.config]
serdever    = 0
root        = "SHINEN_ROOT_PATH"
# /////////////////////////////////////////////////////////////////////////////
# =============================================================================
[admin]
console     = [".*"]
discord-00  = ["DISCORD_USER_ID_00", "DISCORD_USER_ID_01"]
```

### serdever =

設定ファイルには`serdever = number`という項目がいくつか存在します。

serdeverは"serialize desserialize version"の略で、
設定ファイルの構成を管理するバージョンを表します。

**`serdever`は自動的に決定されるので、Aizunaから指示がないかぎり書き換えてはいけません**。

### セクション
#### aizuna

----
##### greeting =
[,greeting][Aizuna コマンド greeting]コマンドによって、Aizunaが表示する文章です。
自由に書き換えて構いません。

----
##### driver =
Aizunaの動作モードを下記の文字列で指定します。デフォルトは`"Thread"`です。

- "Thread": OSスレッドで動作します。
- "Fringe": libfringeを使用してコルーチンで動作します。

詳しくは[Aizuna スレッドとコルーチン][Aizuna スレッドとコルーチン]を参照してください。

----
##### fringe_stack_size =
`"Fringe"`モードで動作する場合のコルーチンのスタックサイズをバイト数で指定します。
デフォルトは`1048576`(= 1024 * 1024 bytes = 1MiB)です。

詳しくは[Aizuna スレッドとコルーチン][Aizuna スレッドとコルーチン]を参照してください。

----
##### path_db =
Aizunaの情報を保持するデータベースへのパスです。デフォルトは`"./db"`です。

相対パスを指定した場合、ルートディレクトリからのパスとして認識されます。
絶対パスを指定することも可能です。

データベースはAizunaによって管理されています。**データベースの内容を書き換えてはいけません**。

----
##### prefix =
[Aizuna コマンド][Aizuna コマンド]を識別するための接頭辞です。デフォルトは`","`です。

```toml
prefix = "aizuna-"
```

このように設定すれば

```
aizuna-help
aizuna-2d6
```

このようにコマンドを使用することが出来ます。

#### [connectors]

外部のチャットサービスと接続するためのコネクタの設定を行います。

コネクタには識別名を決定する必要があります。
自動生成された設定ファイルには`console`と`discord00`が設定されています。

```toml
[connectors.console]
serdever    = 0
enable      = true
connector   = "console"
```

```toml
[connectors.discord-00]
serdever    = 0
enable      = false
connector   = "discord"
[connectors.discord-00.config]
serdever    = 0
token       = "DISCORD_BOT_TOKEN"
```

`[connectors.{識別名}]`のように記述することでコネクタの設定を開始します。

識別名はユーザの判別に使用され、同一のコネクタ識別名を持つユーザ同士のみがセッションを共有できます。
**セッションを作成したコネクタの識別名を書き換えてしまうと、
セッションに接続できなくなってしまう場合があります**。
よってコネクタ識別名の変更は推奨されません。

----
##### enable =

コネクタを使用するかどうかのbool値です。`true`, `false`で指定します。

`false`を指定した場合、そのコネクタは使用されません。

----
##### connector =

コネクタの種別を文字列で指定します。

- `"console"`: デバックコンソールを使用します。
- `"discord"`: Discordを使用します。

> デフォルトでは[connectors.console]が有効、[connectors.discode00]が無効と設定されています。
>
> 実運用環境ではデバックコンソールを無効にし、その他のコネクタを有効化することを推奨します。
>
> Aizunaは全てのコネクタが終了するまで、プロセスを終了しません。[,quit][Aizuna コマンド quit]を参照してください。

----
##### [connectors.{識別名}.config]

コネクタによっては追加の設定が必要となります。

###### "discord"

[Discord 設定][Discord 設定]を参照してください。

#### [rules]

Aizunaで使用するゲームのルールを設定します。

ルールは識別名で区別されます。
自動生成された設定ファイルには`shinen`が設定されています。

```toml
[rules.shinen]
serdever    = 0
enable      = false
prefix      = ","
[rules.shinen.config]
serdever    = 0
root        = "SHINEN_ROOT_PATH"
```

`[rules.{識別名}]`のように記述することでルールの設定を開始します。

----
##### enable =

ルールを使用するかどうかのbool値です。`true`, `false`で指定します。

`false`を指定した場合、そのルールは使用されません。

----
##### prefix =

コマンドでルールを識別する接頭辞を示す文字列です。

----
##### [rules.{識別名}.config]

ルールによっては追加の設定が必要となります。

###### "shinen"

[ShinEn 設定][ShinEn 設定]を参照してください。

#### [admin]

コネクタ毎にAizunaの管理者として扱われるユーザを設定します。

```
{connetor識別名} = ["{Rust正規表現}", "{Rust正規表現}"]
```

デフォルトではconsole接続が管理者として扱われます。

Discord接続はデフォルトでは無効化されています。
有効化する場合、管理者として扱うDiscordID(数字列)を指定してください

```toml
[admin]
console     = [".*"]
discord-00  = ["DISCORD_USER_ID_00", "DISCORD_USER_ID_01"]
```

管理者を設定していないコネクタは、[,quit][Aizuna コマンド quit]を使用できません。
その場合、Aizunaプロセスを強制終了させなくてはいけないことに注意してください。
タイミングによってはデータベースに書き込みが完了していないデータが消失してしまう可能性があります。
**管理者権限を与えるユーザを必ず設定してください**。


{{#include link.md}}


mdbook/src/shinen_config.md
# ShinEn 設定

## 設定ファイル
[Aizuna 設定 ルートディレクトリ][Aizuna 設定 ルートディレクトリ]に存在する`config.toml`に設定を記述する必要があります。

```toml
[rules.shinen]
serdever    = 0
enable      = false
prefix      = ","
[rules.shinen.config]
serdever    = 0
root        = "SHINEN_ROOT_PATH"
```

##### [rules.{識別名}.config]

----
###### root =
ShinEnのファイルを保存するディレクトリへのパスです。

相対パスを指定した場合、Aizunaルートディレクトリからのパスとして認識されます。
絶対パスを指定することも可能です。

## カード
『深淵』はカードを使用しますので、そのデータが必要です。

### カードデータファイル
カードの情報は[ShinEnルートディレクトリ](#root-)に保存されたtoml形式のテキストファイルから読み込まれます。

```shell
{SHINEN_ROOT_PATH}
└── cards
    ├── basic.toml
    └── crimson.toml
```

##### {SHINEN_ROOT_PATH}/cards/basic.toml
基本のカード情報です。

##### {SHINEN_ROOT_PATH}/cards/crimson.toml
『血のごとく赤き 〜夢魔の占い札〜』のカード情報です

### tomlフォーマット

#### 例
```toml
["白の黒剣"]
serdever = 0
name = "白の黒剣"
card_set = "基本"
color = "白"
constellation = "黒剣"
value = 1
story = ["前兆", "出自"]
story_desc = "血が一滴、したたった。無明の闇から、今、何かが誕生し、世界に波紋をもたらしていく。"
action = ["攻撃", "防御"]
damage = "衝撃"
damage_desc = "頭がくらくらする。精神力-4。"
destiny = 1
```

```toml
["狂気"]
serdever = 0
name = "狂気"
card_set = "血のごとく赤き 〜夢魔の占い札〜"
value = 0
desc = "このカードには意味がない。そう意味はない。たぶん……"
story_desc = "いつか お前は私を求めるだろう"
destiny = 95
```

##### ["カード名"]
カード情報の記述を開始します。
##### serdever =
構造のバージョン番号です。現在は0が要求されます。
##### name =
カード名を示す文字列です。
##### card_set =
セットを示す文字列です。
##### color =
色を示す文字列です。
##### constellation =
星座を示す文字列です。
##### value =
カラーナンバーを示す数値です。
##### desc =
カードの文章を示す文字列です。
##### story =
語り部の種別を示す文字列の配列です。
##### story_desc =
語り部の文章を示す文字列です。
##### action =
行動の種別を示す文字列の配列です。
##### damage =
ダメージの種別を示す文字列です。
##### damage_desc =
ダメージの文章を示す文字列です。
##### destiny =
運命番号を示す数値です。


{{#include link.md}}


src/aizuna/config.rs
// -*- mode:rust; coding:utf-8-unix; -*-

//! config.rs

//  Copyright 2017 hanepjiv
//  @author hanepjiv <hanepjiv@gmail.com>
//  @copyright The MIT License (MIT) / Apache License Version 2.0
//  @since 2017/12/28
//  @date 2018/04/28

// ////////////////////////////////////////////////////////////////////////////
// use  =======================================================================
use std::borrow::Cow;
use std::collections::BTreeMap;
use std::fs::File;
use std::io::{Read, Write};
use std::path::{Path, PathBuf};
// ----------------------------------------------------------------------------
use regex::Regex;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
// ----------------------------------------------------------------------------
use super::super::ask::ask;
use super::connector::{Config as ConnectorConfig, Connector, Console, Discord};
use super::rule::shinen::ShinEn;
use super::rule::{Config as RuleConfig, RuleImpl};
use super::{Aizuna, Driver, Error, Result};
// ////////////////////////////////////////////////////////////////////////////
// ============================================================================
const CONFIG_FILE: &str = "config.toml";
const CONFIG_DEFAULT: &str = concat!(
    r##"# -*- mode:toml; coding:utf-8-unix; -*-
# /////////////////////////////////////////////////////////////////////////////
# =============================================================================
# "##,
    env!("CARGO_PKG_NAME"),
    " v",
    env!("CARGO_PKG_VERSION"),
    r##"
serdever                = 0
greeting                = ""##,
    env!("CARGO_PKG_NAME"),
    " v",
    env!("CARGO_PKG_VERSION"),
    r##""
driver                  = "Thread"
fringe_stack_size       = 1048576
path_db                 = "./db"
prefix                  = ","
# /////////////////////////////////////////////////////////////////////////////
# =============================================================================
[connectors.console]
serdever                = 0
enable                  = true
connector               = "console"
# =============================================================================
[connectors.discord-00]
serdever                = 0
enable                  = false
connector               = "discord"
[connectors.discord-00.config]
serdever                = 0
token                   = "DISCORD_BOT_TOKEN"
# /////////////////////////////////////////////////////////////////////////////
# =============================================================================
[rules.shinen]
serdever                = 0
enable                  = false
prefix                  = ","
[rules.shinen.config]
serdever                = 0
root                    = "SHINEN_ROOT_PATH"
# /////////////////////////////////////////////////////////////////////////////
# =============================================================================
[admin]
console                 = [".*"]
discord-00              = ["DISCORD_USER_ID_00", "DISCORD_USER_ID_01"]
"##
);
// ////////////////////////////////////////////////////////////////////////////
// ============================================================================
/// struct Config
#[derive(Debug, Clone)]
pub struct Config {
    /// path_root
    path_root: PathBuf,
    /// path_config
    path_config: PathBuf,
    /// greeting
    greeting: String,
    /// driver
    driver: Driver,
    /// fringe_stack_size
    fringe_stack_size: usize,
    /// path_db
    path_db: PathBuf,
    /// prefix
    prefix: String,
    /// connectors
    connectors: BTreeMap<String, ConnectorConfig>,
    /// rules
    rules: BTreeMap<String, RuleConfig>,
    /// admin
    admin: BTreeMap<String, Vec<Regex>>,
}
// ============================================================================
impl Config {
    // ========================================================================
    /// fn as_greeting
    pub(crate) fn as_greeting(&self) -> &str {
        &self.greeting
    }
    // ========================================================================
    /// fn as_driver
    pub(crate) fn as_driver(&self) -> &Driver {
        &self.driver
    }
    // ========================================================================
    /// fn as_fringe_stack_size
    pub(crate) fn as_fringe_stack_size(&self) -> usize {
        self.fringe_stack_size
    }
    // ========================================================================
    /// fn as_prefix
    pub(crate) fn as_prefix(&self) -> &str {
        self.prefix.as_str()
    }
    // ========================================================================
    /// fn as_path_db
    pub(crate) fn as_path_db(&self) -> &Path {
        &self.path_db
    }
    // ========================================================================
    /// fn as_admin
    pub(crate) fn as_admin(&self) -> &BTreeMap<String, Vec<Regex>> {
        &self.admin
    }
    // ========================================================================
    /// new
    pub fn new<P>(path_root: P) -> Result<Config>
    where
        P: AsRef<Path>,
    {
        info!("Config::new");
        let mut path_root = PathBuf::from(path_root.as_ref());
        path_root = if path_root.is_absolute() {
            path_root
        } else {
            let mut current_dir = ::std::env::current_dir()?;
            current_dir.push(path_root);
            current_dir
        };
        let mut path_config = path_root.clone();
        path_config.push(CONFIG_FILE);
        if !path_config.exists() {
            if ask(
                format!("create config file? {:?}", path_config).as_str(),
                true,
            )? {
                ::std::fs::create_dir_all(path_root.as_path())?;
                let _ = File::create(path_config.as_path()).and_then(
                    |mut f| Ok(f.write(CONFIG_DEFAULT.as_bytes())?),
                )?;
            } else {
                return Err(Error::NoConfig);
            }
        }
        path_root = path_root.canonicalize()?;
        path_config = path_config.canonicalize()?;
        if !path_root.is_dir() {
            return Err(Error::InvalidArg(String::from(
                "input root path is not directory.",
            )));
        }
        let mut config = ::toml::from_str::<Config>(
            File::open(&path_config)
                .and_then(|mut f| {
                    let mut input = String::new();
                    let _ = f.read_to_string(&mut input)?;
                    Ok(input)
                })?
                .as_str(),
        )?;
        if config.path_db.is_relative() {
            let mut path_db = path_root.clone();
            path_db.push(config.path_db);
            config.path_db = path_db;
        }
        Ok(Config {
            path_root,
            path_config,
            ..config
        })
    }
    // ========================================================================
    /// aizuna
    pub fn aizuna(self) -> Result<Aizuna> {
        info!(
            "::aizuna::Config::aizuna: {:?}",
            self.path_config
        );
        let mut connectors = Vec::<Box<dyn Connector>>::default();
        for (k0, v0) in &self.connectors {
            if !v0.enable {
                continue;
            }
            match v0.connector.as_str() {
                "console" => {
                    connectors.push(Box::new(Console::new(k0)));
                }
                "discord" => {
                    connectors
                        .push(Box::new(Discord::new(k0, v0.config.clone())?));
                }
                _ => {
                    return Err(Error::InvalidArg(format!(
                        "::aizuna::Config::aizuna: unknown connector: {}",
                        v0.connector
                    )));
                }
            }
        }
        let mut rules = BTreeMap::<String, RuleImpl>::default();
        for (k0, v0) in &self.rules {
            if !v0.enable {
                continue;
            }
            match k0.as_str() {
                "shinen" => {
                    let _ = rules.insert(
                        v0.prefix.clone(),
                        RuleImpl::ShinEn(ShinEn::new(
                            self.path_root.as_path(),
                            v0.config.clone(),
                        )?),
                    );
                }
                _ => {
                    return Err(Error::InvalidArg(format!(
                        "::aizuna::Config::aizuna: unknown rule: {}",
                        k0
                    )));
                }
            }
        }
        Aizuna::new(self, connectors, rules)
    }
}
// ============================================================================
impl Serialize for Config {
    fn serialize<S>(
        &self,
        serializer: S,
    ) -> ::std::result::Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self::serialize::Config::from(self).serialize(serializer)
    }
}
// ============================================================================
impl<'de> Deserialize<'de> for Config {
    fn deserialize<D>(deserializer: D) -> ::std::result::Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        self::serialize::Config::deserialize(deserializer)?
            .into()
            .map_err(|e| ::serde::de::Error::custom(format!("{}", e)))
    }
}
// ////////////////////////////////////////////////////////////////////////////
// ============================================================================
/// serialize
mod serialize {
    // ////////////////////////////////////////////////////////////////////////
    // ========================================================================
    use super::*;
    // ////////////////////////////////////////////////////////////////////////
    // ========================================================================
    const CURRENT: i32 = 0i32;
    const AGE: i32 = 0i32;
    // ////////////////////////////////////////////////////////////////////////
    // ========================================================================
    /// struct Config
    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub(crate) struct Config<'a> {
        /// serdever
        serdever: i32,
        /// greeting
        greeting: Option<Cow<'a, str>>,
        /// driver
        driver: Option<Cow<'a, Driver>>,
        /// fringe_stack_size
        fringe_stack_size: Option<usize>,
        /// path_db
        path_db: Option<Cow<'a, str>>,
        /// prefix
        prefix: Option<Cow<'a, str>>,
        /// connectors
        connectors: Option<Cow<'a, BTreeMap<String, ConnectorConfig>>>,
        /// rules
        rules: Option<Cow<'a, BTreeMap<String, RuleConfig>>>,
        /// admin
        admin: Option<BTreeMap<String, Vec<Cow<'a, str>>>>,
    }
    // ========================================================================
    impl<'a> Config<'a> {
        // ====================================================================
        /// into
        pub(crate) fn into(self) -> Result<super::Config> {
            debug!("::aizuna::Config::serialize::into");
            if self.serdever < (CURRENT - AGE) || CURRENT < self.serdever {
                return Err(Error::SerDeVer(self.serdever, CURRENT, AGE));
            }
            match self.serdever {
                0 => Ok(super::Config {
                    path_root: PathBuf::default(),
                    path_config: PathBuf::default(),
                    greeting: self.greeting.map_or(
                        String::from(concat!(
                            env!("CARGO_PKG_NAME"),
                            " v",
                            env!("CARGO_PKG_VERSION"),
                        )),
                        String::from,
                    ),
                    driver: self.driver
                        .map_or(Driver::Thread, Cow::into_owned),
                    fringe_stack_size: self.fringe_stack_size
                        .unwrap_or(1024 * 1024),
                    path_db: self.path_db
                        .map(Cow::into_owned)
                        .map_or(PathBuf::from("./db"), PathBuf::from),
                    prefix: self.prefix
                        .map_or(String::from(","), String::from),
                    connectors: self.connectors
                        .map(Cow::into_owned)
                        .unwrap_or_default(),
                    rules: self.rules
                        .map(Cow::into_owned)
                        .unwrap_or_default(),
                    admin: if let Some(x) = self.admin {
                        let mut ret = BTreeMap::default();
                        for (k, v) in x {
                            let mut vs = Vec::default();
                            for i in v {
                                vs.push(Regex::new(&i.into_owned())?);
                            }
                            let _ = ret.insert(k, vs);
                        }
                        ret
                    } else {
                        BTreeMap::default()
                    },
                }),
                _ => Err(Error::SerDeVer(self.serdever, CURRENT, AGE)),
            }
        }
    }
    // ========================================================================
    impl<'a> From<&'a super::Config> for Config<'a> {
        fn from(src: &'a super::Config) -> Self {
            debug!("::aizuna::Config::serialize::from");
            Self {
                serdever: CURRENT,
                greeting: Some(From::from(src.greeting.as_str())),
                driver: Some(Cow::Borrowed(&src.driver)),
                fringe_stack_size: Some(src.fringe_stack_size),
                path_db: Some(src.path_db.as_os_str().to_string_lossy()),
                prefix: Some(From::from(src.prefix.as_str())),
                connectors: Some(Cow::Borrowed(&src.connectors)),
                rules: Some(Cow::Borrowed(&src.rules)),
                admin: if src.admin.is_empty() {
                    None
                } else {
                    let mut ret = BTreeMap::default();
                    for (k, v) in &src.admin {
                        let mut vs = Vec::default();
                        for i in v.iter() {
                            vs.push(From::from(i.as_str()));
                        }
                        let _ = ret.insert(k.clone(), vs);
                    }
                    Some(ret)
                },
            }
        }
    }
}
